### PR TITLE
Bump `coverage.py` to v7.3.0 in `ansible-test`

### LINKED ISSUE #80427: Update the supported coverage versions in ansible-test
### Summary

Add support for the next version of `coverage` as needed to support Python 3.12.

### Issue Type

Feature Idea

### Component Name

`ansible-test`

### PR DESCRIPTION
**Pull request by webknjaz**
This patch upgrades the `coverage.py` version used under Python 3.8 through 3.12.
It also upgrades the path patterns in the generated configs according to the changes
in v7.0.0b1 of said library as it replaced the matching method from `fnmatch` to `glob`[[1]][[2]].

Resolves #80427

[1]: https://coverage.readthedocs.io/en/7.2.7/migrating.html#migrating-to-coverage-py-7-x
[2]: https://github.com/nedbat/coveragepy/commit/ec6205a8de972af6a09453235d02a7ebea6aea8e

##### SUMMARY

^ $sbj ^

##### ISSUE TYPE

- Maintenance Pull Request
- Test Pull Request

##### COMPONENT NAME

`ansible-test`

##### ADDITIONAL INFORMATION

N/A

### CODE REVIEW

**Review by mattclay — COMMENTED** [review]
@webknjaz The changes *looked* correct, but I decided to test the glob matching to be sure. At first I tried reverting the glob portion of your changes in this PR -- yet code coverage was still reported. Then I tried using the API directly and I'm seeing the old behavior:

```python
from coverage.files import GlobMatcher

print(GlobMatcher('/hello/*').match('/hello/too/deep/to/match.py'))
```

I'd expect that *not* to match, since I didn't use `**`, but it does. Can you confirm that glob matching is (or is not) working as documented?

**Comment by mattclay** [comment]
Also, once we've merged this PR with whatever coverage version we pick, we'll want to update the spare-tire repo:

https://github.com/ansible/spare-tire/blob/b23b16c077ce0e8ada65c12b79a608025d4fb1fd/wheel_matrix.yml#L94

We should be able to just replace `6.5.0` with the new pinned version. The purpose of that is to ensure that if we add a new platform/version later, that it will get a wheel build for the pinned version (not just latest). We shouldn't need to keep the old entry around though, since we won't be adding any new platform/version combinations using the old version.

**Comment by webknjaz** [comment]
It looks like `/*` at the very end of a pattern is a corner case, specifically defined here: https://github.com/nedbat/coveragepy/commit/ec6205a8de972af6a09453235d02a7ebea6aea8e#diff-fd153710f7f1f017fbf6798fb808fc2852d59ccab90f6393e010d740d30ef804R292.
In this regard, it does not behave like https://docs.python.org/3/library/glob.html.
```pycon
In [31]: GlobMatcher(('/hello/*/thing/', )).re
Out[31]: re.compile(r'(?:[/\\]hello[/\\][^/\\]*[/\\]thing[/\\])\Z', re.UNICODE)

In [32]: GlobMatcher(('/hello/*/', )).re
Out[32]: re.compile(r'(?:[/\\]hello[/\\][^/\\]*[/\\])\Z', re.UNICODE)

In [33]: GlobMatcher(('/hello/*', )).re
Out[33]: re.compile(r'(?:[/\\]hello[/\\].*)\Z', re.UNICODE)

In [34]: GlobMatcher(('/hello/*/thing/', )).match('/hello/too/deep/to/match.py')
Out[34]: False

In [35]: GlobMatcher(('/hello/*/too/', )).match('/hello/too/deep/to/match.py')
Out[35]: False

In [36]: GlobMatcher(('/hello/*/deep/', )).match('/hello/too/deep/to/match.py')
Out[36]: False

In [37]: GlobMatcher(('/hello/*/deep/*', )).match('/hello/too/deep/to/match.py')
Out[37]: True

In [38]: GlobMatcher(('/hello/*/', )).match('/hello/too/deep/to/match.py')
Out[38]: False

In [39]: GlobMatcher(('/hello/*', )).match('/hello/too/deep/to/match.py')
Out[39]: True

In [40]: GlobMatcher(('/hello/*', )).match('/hello/too/')
Out[40]: True

In [41]: GlobMatcher(('/hello/*/', )).match('/hello/too/')
Out[41]: True

In [42]: GlobMatcher(('/hello/*/', )).match('/hello/too/deep')
Out[42]: False

In [43]: GlobMatcher(('/hello/*/', )).match('/hello/too/deep/')
Out[43]: False
```

It is unclear why this specific corner case was added, as it is not documented explicitly, nor does the commit add any tests for this case. It might be intended or a bug, there's no way to deduce it w/o asking..

**Comment by webknjaz** [comment]
@mattclay I submitted https://github.com/nedbat/coveragepy/pull/1650 — let's see if that gets us the desired behavior in the end.

File: test/lib/ansible_test/_internal/coverage_util.py [hunk 1]
```diff
@@ -69,7 +69,7 @@ class CoverageVersion:
 
 COVERAGE_VERSIONS = (
     # IMPORTANT: Keep this in sync with the ansible-test.txt requirements file.
-    CoverageVersion('6.5.0', 7, (3, 7), (3, 11)),
+    CoverageVersion('7.2.7', 7, (3, 7), (3, 12)),
```
Comment (mattclay) [inline test/lib/ansible_test/_internal/coverage_util.py]: Now that 7.3.0 is out, we should use that. Since it drops Python 3.7 support we'll need a new entry here:

```suggestion
    CoverageVersion('7.3.0', 7, (3, 8), (3, 12)),
    CoverageVersion('6.5.0', 7, (3, 7), (3, 7)),
```

File: test/lib/ansible_test/_data/requirements/ansible-test.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
 # The test-constraints sanity test verifies this file, but changes must be made manually to keep it in up-to-date.
 virtualenv == 16.7.12 ; python_version < '3'
-coverage == 6.5.0 ; python_version >= '3.7' and python_version <= '3.11'
+coverage == 7.2.7 ; python_version >= '3.7' and python_version <= '3.12'
```
Comment (mattclay) [inline test/lib/ansible_test/_data/requirements/ansible-test.txt]: Now that 7.3.0 is out, we should use that. Since it drops Python 3.7 support we'll need a new entry here:

```suggestion
coverage == 7.3.0 ; python_version >= '3.8' and python_version <= '3.12'
coverage == 6.5.0 ; python_version >= '3.7' and python_version <= '3.7'
```

**Review by mattclay — DISMISSED** [review]
Since there hasn't been any movement on https://github.com/nedbat/coveragepy/pull/1650, I think we should move on without waiting for that. That appears to be safe to do, since the only issue seems to be that `/*` counts as `/**`, correct?

**Comment by webknjaz** [comment]
> That appears to be safe to do, since the only issue seems to be that `/*` counts as `/**`, correct?

Yes, and I think that sticking `**` everywhere like I've done in this patch is the way to go and doesn't need any special-casing, right?

**Review by mattclay — COMMENTED** [review]
Looks good. It just needs a changelog fragment.

**Comment by webknjaz** [comment]
> just needs a changelog fragment.

Interestingly, I haven't found any changelog fragments coupled with the previous version bumps — I was hoping for an example of what could be expected.

File: test/lib/ansible_test/_internal/coverage_util.py [hunk 3]
```diff
@@ -254,13 +255,13 @@ def generate_ansible_coverage_config() -> str:
...
 
 omit =
-    */python*/dist-packages/*
-    */python*/site-packages/*
-    */python*/distutils/*
-    */pyshared/*
-    */pytest
-    */AnsiballZ_*.py
-    */test/results/*
+    **/python*/dist-packages/**
```
Comment (webknjaz) [inline test/lib/ansible_test/_internal/coverage_util.py]: @mattclay should we revert these after all?

Comment (mattclay) [reply test/lib/ansible_test/_internal/coverage_util.py]: Just to confirm, the new `**` behavior is only for matching in the middle of a path? The `*/` and `/*` matches at the start and end of a path continue to work without `**`?

If that's the case, then yes, we should revert the path changes, since it seems the coverage docs will be updated, rather than the implementation.

Comment (webknjaz) [reply test/lib/ansible_test/_internal/coverage_util.py]: Yes, that's correct. I've been thinking of documenting this upstream, since it's unlikely that Ned will accept the original change I offered.

File: changelogs/fragments/ansible-test-coverage-update.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+
+bugfixes:
```
Comment (webknjaz) [inline changelogs/fragments/ansible-test-coverage-update.yml]: This should probably be
```suggestion
minor_changes:
```

**Review by mattclay — APPROVED** [review]
